Add Customer#deleted?

iloveitaly opened the pull request:
A ruby-like method to easily query the `deleted` status of a `Stripe::Customer`

rodrigoargumedo commented:
+1

brandur commented on lib/stripe/customer.rb:
```diff
@@ -54,6 +54,10 @@ def delete_discount
     end
 
+    def deleted?
+      self.respond_to?(:deleted)
```
If I'm reading this correctly, it looks like you might just be checking whether the object responds to a call to `deleted`. This is good, but you'd probably also want to check the value inside `#deleted` as well.

brandur commented:
@iloveitaly Thanks a lot for the contribution and our apologies for the slow turnaround in getting back to you. This looks great but we also have one other open pull request that's going a similar thing. Would you mind taking a look at #274 and confirming that it would also be an adequate alternative for you?

iloveitaly commented:
#274 would definitely cover this need! I'll hold off on any improvements here unless that PR can't be merged in. Thanks for taking a look at this!

brandur commented:
@iloveitaly Cool! Thanks for chiming in. We'll see what we can do about #274.

brandur commented:
Closing this in favor of #309. Thank-you for sending this in though!

iloveitaly commented:
Hey @brandur! Unfortunately, #309 doesn't fix this use case :(

`Stripe::Customer` only contains `deleted = true` when the customer is deleted. If the customer is not deleted then, the customer does not contain that field at all so `Customer#deleted?` throws a no method error.

I'm seeing two paths forward here: we add an explicit `deleted?` method on `Stripe::Customer` or modify the Stripe API to return `deleted = false` with a non-deleted customer.

kyleconroy commented:
I don't think we want to add `deleted = false` to all customers, so I'm fine for adding an explicit `Customer#deleted?`.

brandur commented:
> I don't think we want to add deleted = false to all customers, so I'm fine for adding an explicit Customer#deleted?.

Cool. FWIW, I tend to think that stable API responses (i.e. same set of fields regardless of what state they're in) are generally a good idea, but looking through the API code it seems that doing this might be a bit of a yak shave because theoretically any "paranoid" resource could be returned as deleted and we'd probably want to go through and document them all.

Regardless of that though, I also think an explicit helper might be a good idea here because it's helpful and isn't going to be detrimental any longer term design that we might want to do. I'll add one in. I'll probably at it at the level of `StripeObject` because we can have different types of deleted objects.

kyleconroy commented:
> I'll probably at it at the level of StripeObject because we can have different types of deleted objects.

`Plan` and `Subscription` objects can also be deleted, so this makes sense.

iloveitaly commented:
> Plan and Subscription objects can also be deleted, so this makes sense.

From what I can tell, you can't grab a reference to a plan/subscription directly after they are deleted. When these deleted records are referenced (a deleted plan in an `Invoice#lines`) they don't contain `deleted = true`.

Might be easier to explain in code:

```
customer = create_customer_with_card

plan = Stripe::Plan.create({
  :amount => 10_00,
  :interval => 'month',
  :name => 'A Plan',
  :currency => 'usd',
  :id => 'plan_CMFNSWINCYSGXJF'
})

sub = customer.subscriptions.create(plan: plan.id)
plan.delete

# at this point `plan.deleted == true`

# this will raise an exception
fresh_plan = Stripe::Plan.retrieve(plan.id)

customer.refresh

plan_line = customer.invoices.first.lines.first.plan
# plan_line.deleted will throw a no method error

customer.subscriptions.delete

# `customer.subscriptions.retrieve(sub.id)` throws a `Stripe::InvalidRequestError` error
```

It would be great if either:

* `customer.invoices.first.lines.first.plan.deleted == true` when the plan was deleted
* deleted plans worked in a similar way to deleted customers, i.e. a deleted plan could be `retrieve`d and `plan.deleted == true`

Right now the only way to determine if a `customer.invoices.first.lines.first.plan.deleted` is to attempt to save/retrieve the plan and inspect the exception's error message.
